src/differentials.jl

```diff
@@ -272,6 +282,74 @@ function itself, when that function is not a closure.
 """
 const NO_FIELDS = DoesNotExist()
 
+
+"""
+    Composite{P, T} <: AbstractDifferential
+
+This type represents the differential for a `struct`/`NamedTuple`, or `Tuple`.
+`P` is the the corresponding primal type that this is a differential for.
+
+`Composite{P}` should have fields (technically properties), that match to a subset of the
+fields of the primal type; and each should be a differential type matching to the primal
+type of that field.
+Fields of the P that are not present in the Composite are treated as `Zero`.
+
+`T` is an implementation detail representing the backing data structure.
+For Tuple it will be a Tuple, and for everything else it will be a `NamedTuple`.
+It should not be passed in by user.
+"""
+struct Composite{P, T} <: AbstractDifferential
+    backing::T
+end
+# Note: If T is a Tuple, then P is also a Tuple
+# (but potentially a different one, as it doesn't contain differentials)
+
+function Composite{P}(; kwargs...) where P
+    backing = (; kwargs...)  # construct as NamedTuple
+    return Composite{P, typeof(backing)}(backing)
+end
+
+function Composite{P}(args...) where P
+    return Composite{P, typeof(args)}(args)
+end
+
+function Base.show(io::IO, comp::Composite{P}) where P
+    print(io, "Composite{")
+    show(io, P)
+    print(io, "}")
+    # allow Tuple or NamedTuple `show` to do the rendering of brackets etc
+    show(io, backing(comp))
+end
+
+Base.convert(::Type{<:NamedTuple}, comp::Composite{<:Any, <:NamedTuple}) = backing(comp)
+Base.convert(::Type{<:Tuple}, comp::Composite{<:Any, <:Tuple}) = backing(comp)
+
+Base.getindex(comp::Composite, idx) = getindex(backing(comp), idx)
+Base.getproperty(comp::Composite, idx::Int) = getproperty(backing(comp), idx)  # for Tuple
+Base.getproperty(comp::Composite, idx::Symbol) = getproperty(backing(comp), idx)
+Base.propertynames(comp::Composite) = propertynames(backing(comp))
+
+Base.iterate(comp::Composite, args...) = iterate(backing(comp), args...)
+Base.length(comp::Composite) = length(backing(comp))
+Base.eltype(::Type{<:Composite{<:Any, T}}) where T = eltype(T)
+
+function Base.map(f, comp::Composite{P, <:Tuple}) where P
+    vals::Tuple = map(f, backing(comp))
+    return Composite{P, typeof(vals)}(vals)
+end
+function Base.map(f, comp::Composite{P, <:NamedTuple{L}}) where{P, L}
+    vals = map(f, Tuple(backing(comp)))
+    named_vals = NamedTuple{L, typeof(vals)}(vals)
+    return Composite{P, typeof(named_vals)}(named_vals)
+end
+
+Base.conj(comp::Composite) = map(conj, comp)
+
+extern(comp::Composite) = backing(map(extern, comp))  # gives a NamedTuple or Tuple
+
+
+#==============================================================================#
```

Comment:
More excess whitespace, and unnecessary comment lines.

I feel like if we're doing this to mark off `sections`, we should just be moving that code into its own file.

Reply:
seems legit, will do in a seperate PR to avoid big diffs